src/event.rs

```diff
@@ -74,9 +74,105 @@ pub enum Event<'a, T: 'static> {
     UserEvent(T),
 
     /// Emitted when the application has been suspended.
+    ///
+    /// # Portability
+    ///
+    /// Not all platforms support the notion of suspending applications, and there may be no
+    /// technical way to guarantee being able to emit a `Suspended` event if the OS has
+    /// no formal application lifecycle (currently only Android and iOS do). For this reason
+    /// Winit does not currently try to emit pseudo `Suspended` events before the application
+    /// quits on platforms without an application lifecycle.
+    ///
+    /// Considering that the implementation of `Suspended` and [`Resumed`] events may be internally
+    /// driven by multiple platform-specific events, and there may be subtle differences across
+    /// platforms with how these internal events are delivered, it's recommended that applications
+    /// be able to gracefully handle redundant (i.e. back-to-back) `Suspended` or [`Resumed`] events.
+    ///
+    /// Also see [`Resumed`] notes.
+    ///
+    /// ## Android
+    ///
+    /// On Android specifically, the `Suspended` event relates to two things:
+    /// 1. The application is paused (according to the [Activity lifecycle])
+    /// 2. Or the application has had its associated [`SurfaceView`] destroyed (which is expected
+    /// to closely correlate with being paused)
+    ///
+    /// Applications that need to run on Android should assume their [`SurfaceView`] has been
+    /// destroyed, which indirectly invalidates any existing render surfaces that may have been
+    /// created outside of Winit (such as an `EGLSurface`, [`VkSurfaceKHR`] or [`wgpu::Surface`]).
```

Comment:
As discussed before I'm not fond of this wording: you first state that it could be _either_ destroyed or just paused, then continue stating that users should assume it has been destroyed. Is it perhaps better to state that it has been destroyed first, followed by mentioning that "this is likely/typically because the application has been paused (according to the [Activity lifecycle])"?

Reply:
The intent here is to let people know how the implementation currently works, which might be important if you're dealing with some Android-specific use case.

The use of "Or" is because that's how the actual implementation work right now, which isn't being changed. There are two things that may lead result in getting a "Suspended" event on Android, and you don't need both (so it's "Or" not "And").

The second part is effectively saying that developers should assume the worst. Since there are two possible causes for the app to be suspended and one of them technically doesn't invalidate their surface view then I think it's best to recommend that applications should behave as if their surface view has been destroyed.

I.e. the first part says exactly how the implementation works; the second part recommends that you always treat it like your surface view was destroyed.

Personally I think the information makes sense and is consistent with itself, so I'm not quite sure how to improve the wording.

Your suggestion seems to effectively remove the information about how it works internally which seems like it could be useful to know for some Android specific use cases.

If I change it to "state that it has been destroyed first" that's not always true, but I do recommend that you act as if it's always true.

Reply:
> The use of "Or" is because that's how the actual implementation work right now, which isn't being changed

It feels like we've discussed this before, but I've forgot the outcome, so I'll ask again [^1]:

[^1]: and to restate: long walls of text repeating the same thing over and over make this only harder to discuss.

Where exactly is Android's `onPause` callback - resulting in `Event:Pause` - converted into `Event::Suspended` on the Winit side? All I can see - when checked out on this PR - is solely `Event::WindowDestroyed` resulting in `Event::Suspended`. Furthermore, digging down for calls to that callback show only `onSurfaceChanged`/`onSurfaceDestroyed` calls on `android::view::Surface`.

https://cs.android.com/android/platform/superproject/+/master:frameworks/native/include/android/native_activity.h;l=204;bpv=1;bpt=1?q=onNativeWindowDestroyed

Reply:
Ah, damn, I think I'm recalling how my android-activity branch works, not master, urgh sorry :/

I think the way I described it is how it _should_ be implemented but right now that's not technically the case.

I guess I'll just drop the Android docs for now and instead this info can be added if the implementation is updated (fixed imho).

Reply:
okey, I've updated the technical descriptions of how the events are implemented currently to say:

> On Android, the `Suspended` event is only sent when the application's associated
> [`SurfaceView`] is destroyed. This is expected to closely correlate with the [`onPause`]
> lifecycle event but there may technically be a discrepancy.

and:
>  On Android the `Resumed` event is sent when a new [`SurfaceView`] has been created. This is
> expected to closely correlate with the [`onResume`] lifecycle event but there may technically
> be a discrepancy.

Reply:
@rib Thanks, that's what I was trying to get across all along. The documentation in its current form is exactly what I had been looking for, good job on that.  I completely agree with ``This is expected to closely correlate with the [`onPause`] lifecycle event but there may technically be a discrepancy.`` as well.

---

Whether or not it is a good idea to generalize these in the future is something we should continue to discuss in https://github.com/rust-windowing/winit/issues/2337. I have some extra feelings on the `OR` and `AND` case for `Suspended` and `Resumed` that I'll share with you there.